Fixes last causes for WSOD

MorrisJobke opened the pull request:
- print plain text if rendering of exception/error page is not possible

Steps to test:
- change version.php to a lower version number ;)

This will make if easier to spot the problems without digging into owncloud.log. It also now properly logs the exception instead of just dumping the PHP "uncaught exception" text.

MorrisJobke committed:
Fixes last causes for WSOD

* print plain text if rendering of exception/error page is not possible

MorrisJobke commented:
https://github.com/owncloud/core/pull/19537/files?w=1 <- makes it easier to understand what I did.


@LukasReschke @nickvergessen @PVince81 @DeepDiver1975 Please review :)

LukasReschke commented on index.php:
```diff
@@ -46,12 +46,27 @@
+		OC_Template::printExceptionErrorPage($ex);
+	} catch (Exception $exNew) {
+		\OCP\Util::logException('index', $exNew);
+		print($ex->getMessage());
```
Mhm. `printExceptionErrorPage` would not leak any information as it checks for debug mode. If we do this here then we might leak information. Can we adjust this somehow?

DeepDiver1975 commented on index.php:
```diff
@@ -46,12 +46,27 @@
+	try {
+		OC_Response::setStatus(OC_Response::STATUS_SERVICE_UNAVAILABLE);
+		OC_Template::printErrorPage($ex->getMessage(), $ex->getHint());
+	} catch (Exception $exNew) {
```
what about adding this try catch logic to printErrorPage() and printExceptionErrorPage()

BTW: why do we have two almost similar function?

MorrisJobke replied:
> BTW: why do we have two almost similar function?

Ask @LukasReschke

LukasReschke replied:
`printErrorPage` is used to display an error page with a developer defined error message. Such as the HintException where we show something like "Please install MySQL".

`printExceptionErrorPage` will instead just show a generic error page and some debug information if debug mode is enabled. There is no way to define a custom error message.

Thinking about that, what about killing  `printErrorPage` and adding an additional message parameter to `printExceptionErrorPage`?

DeepDiver1975 replied:
> Thinking about that, what about killing printErrorPage and adding an additional message parameter to printExceptionErrorPage?

:+1:

MorrisJobke replied:
Just looking at the code and printErrorPage is used also in OC_Files, OC::init(), OC::checkConfig, remote.php and public.php.

It's not that easy to kill and then we just combine the functionality of both in one method with two code paths... for now I will add the exception handling to both methods to have it covered properly, but I will not refactor it now (because we will release soon).

MorrisJobke commented:
@DeepDiver1975 @LukasReschke I added the handling now to the OC_Template. I also reduced the internal dependencies to a minimum (only the logger) and used then just `header()` and `print()` to actually get the text printed.

https://github.com/owncloud/core/pull/19537/files?w=1

@LukasReschke I also addressed your comment and print the default text for unexpected exceptions.

LukasReschke commented on lib/private/template.php:
```diff
@@ -300,10 +300,20 @@ public static function printGuestPage( $application, $name, $parameters = array(
+			$logger->error("$error_msg $hint", ['app' => 'core']);
+			$logger->logException($e, ['app' => 'core']);
+
+			header($_SERVER['SERVER_PROTOCOL'] . ' 500 Internal Server Error');
```
https://github.com/owncloud/security-tracker/issues/119 - THX

MorrisJobke replied:
Way to fix this? Check for HTP/2, HTTP/1.1 or HTTP/1.0?

LukasReschke replied:
I'd add a helper to IRequest `getHttpProtocol()` that does this magic. - Then we can in a second step replace it everywhere else as well.

MorrisJobke replied:
As you know I want to introduce as few dependencies as possible here - so I would do this here in place to allow proper error print.

LukasReschke replied:
K. - Maybe we should make IRequest have less dependencies in the future :see_no_evil:

LukasReschke replied:
https://github.com/owncloud/core/pull/19605 - for all the other occurences.

MorrisJobke replied:
Fixed.

LukasReschke commented on lib/private/template.php:
```diff
@@ -312,18 +322,31 @@ public static function printErrorPage( $error_msg, $hint = '' ) {
+			$logger->logException($exception, ['app' => 'core']);
+			$logger->logException($e, ['app' => 'core']);
+
+			header($_SERVER['SERVER_PROTOCOL'] . ' 500 Internal Server Error');
```
As well

MorrisJobke committed:
validate HTTP protocol in case of an exception

MorrisJobke commented:
@LukasReschke @nickvergessen @DeepDiver1975 Ready for review: https://github.com/owncloud/core/pull/19537/files?w=1

nickvergessen commented on lib/private/template.php:
```diff
@@ -312,22 +322,57 @@ public static function printErrorPage( $error_msg, $hint = '' ) {
+	 * @return string HTTP protocol. HTTP/2, HTTP/1.1 or HTTP/1.0.
+	 * @internal Don't use this - use AppFramework\Http\Request->getHttpProtocol instead
+	 */
+	public static function getHttpProtocol() {
```
Why public

MorrisJobke replied:
Because otherwise it can't be access statically? Or is this possible with self:: and protected?

MorrisJobke replied:
I tested and made it protected. It still works

nickvergessen commented on lib/private/template.php:
```diff
@@ -300,10 +300,20 @@ public static function printGuestPage( $application, $name, $parameters = array(
+			$logger->error("$error_msg $hint", ['app' => 'core']);
+			$logger->logException($e, ['app' => 'core']);
+
+			header(self::getHttpProtocol() . ' 500 Internal Server Error');
```
Use `IRequest::getHttpProtocol()`

MorrisJobke replied:
No. Because in this case I want to have as few dependencies as possible, because an exception was thrown and the templating itself also has thrown an exception. It is very likely that fetching yet another object from the DI container will cause also an exception and we are lost. Also IRequest is quite heavy on it's own dependencies.

nickvergessen replied:
ok

LukasReschke commented:
:+1:

MorrisJobke commented:
@nickvergessen Mind to review?

nickvergessen committed:
Do not trigger the update exception when we already show an error

nickvergessen commented:
> change version.php to a lower version number ;)

I wonder why this works on testing, because when we implemented the downgrade blocker it showed the normal blue page.

Okay, found the reason, easy to resolve.

:+1: for your code

MorrisJobke commented:
> Okay, found the reason, easy to resolve.

Thanks - tested and works :+1:
